src/mongocxx/private/collection.hpp

```diff
@@ -39,6 +39,13 @@ class collection::impl {
           client_impl(client) {
     }
 
+    impl(const impl& i) : collection_t{libmongoc::collection_copy(i.collection_t)} { }
```

Comment:
This does bring up a larger issue @acmorrow, that our memory management is not exactly harmonized with the standard c++ facilities like `set_new_handler`, etc. Perhaps we want to use `bson_mem_set_vtable` so that `bson_malloc` uses `operator new` ?